### ISSUE TITLE
sign addons with autograph /sign/file

### ISSUE THREAD

**Issue by g-k** [mozilla/addons-server#9308]
Per https://github.com/mozilla/addons-server/pull/9306#issuecomment-417300984, we want to switch addons-server to the autograph [`/sign/file` endpoint](https://github.com/mozilla-services/autograph/tree/6bc03e3ce0ce8aae6e31b8240159ec2c7bceef86/signer/xpi#signature-request).

This will let us eventually migrate to SHA256 digests for the PKCS7 signature and add COSE Signatures for https://bugzilla.mozilla.org/show_bug.cgi?id=1471186

We'll probably want to:

- [ ] QA the autograph-signed addons (this probably entails fixing ~~https://github.com/mozilla-services/autograph/issues/130~~ (edit: not necessary; closed) and https://github.com/mozilla-services/autograph/issues/122 and making sure autograph XPIs pass other signing-clients and addons-server unit tests)  
- [ ] determine peak load and network capacity and load test autograph

**Comment by muffinresearch** [comment]
@g-k The google doc detailing the timeline has become a bit unwieldy. I want to make sure I get a better understanding of the steps involved from the AMO perspective, including when we need to make these changes and what we need to know about how to test this.

Would you be able to detail the implications of this change (wrt backwards compatibility and versions impacted) and where it fits into the current timeline?

**Comment by g-k** [comment]
@muffinresearch Sure, the autograph defaults match the current signature format (PKCS7 signature with SHA1 digest) and I'd expect any issues to be related to differences in packaging and load due to sending whole XPIs to autograph.

After switching to `/sign/file`, we can set [/sign/file request params](https://github.com/mozilla-services/autograph/blob/master/signer/xpi/README.rst#signature-request) to use stronger PKCS7 signature digests and add COSE signatures (probably a separate issue).

The goal is for Fx68 to require addons have a PKCS7 signature with a SHA256 digest and COSE signature for Mid-2019. For AMO that probably means re-signing old addons with the new signature and maintaining two copies of each addon w/ each signature type for awhile.

Also, if QA determines that Fx57 ignores COSE signatures and accepts the PKCS7 signature with a SHA1 digest, then we can start issuing backwards compatible COSE signatures from AMO right away and reduce the number of addons to re-sign later.

**Comment by muffinresearch** [comment]
@g-k my understanding is that we should wait on the outcome of https://bugzilla.mozilla.org/show_bug.cgi?id=1477701 and then if that looks good that should green-light moving to use the `/sign/file` endpoint. 

Does that sounds right to you? I think Q4 should be ok assuming we know we're good to proceed.

It sounded to me talking to shell, that if it works incorporating the multiple signing mechanisms without compatibility issues  for older Firefox versions we might avoid having to have 2 differently signed versions of each add-on.

**Comment by g-k** [comment]
Yeah, that makes sense. There's no need to move to the endpoint before we
know what params to use.

**Comment by g-k** [comment]
OK heard back from QA today. Add-on variant A params (PKCS7 SHA1 and COSE ES256) work in Fx <57, so we can switch to that without breaking backwards compatibility.

**Comment by g-k** [comment]
I forgot to update and close https://bugzilla.mozilla.org/show_bug.cgi?id=1477701, but none of the other issues referenced here are blockers.

We did some XPI singing load testing in https://bugzilla.mozilla.org/show_bug.cgi?id=1508400

**Comment by EnTeQuAk** [comment]
@g-k @jvehent so from a very naive query the largest webextension add-ons that we have are ~114-197MB which is the absolute top. Since we're denying uploads > 100MB these days though I think we should be fine.

**Comment by EnTeQuAk** [comment]
> That won't be an issue. Autograph's post limit is currently set to 1GB. mozilla-services/autograph:handlers.go@master#L75-L79

Perfect, just wanted to verify. Thanks!

**Comment by EnTeQuAk** [comment]
QA: This is not yet enabled on -dev and -stage. I'll send around a separate email for how and when we're going to test this.

**Comment by AlexandraMoga** [comment]
@EnTeQuAk Thanks for the email you sent the other day which brings more information about this implementation.
While I did understand that the switch will be turned on for -dev on Monday, I still have some unclarities with regards to "how" we are going to test this.

For example, are there any changes to be observed in the `META-INF` to identify this new signature?

**Comment by g-k** [comment]
> For example, are there any changes to be observed in the META-INF to identify this new signature?

@AlexandraMoga yep! Addons signed with this format will have `META-INF/cose.manifest` and `META-INF/cose.sig` files in addition to the other signature files (`manifest.mf`, `mozilla.sf`, and `mozilla.rsa`).

**Comment by EnTeQuAk** [comment]
In addition to that, if I recall correctly, there shouldn't be any `MD5` hashes anymore in `META-INF/manifest.mf`, only `SHA1` and `SHA256` hashes.

**Comment by g-k** [comment]
> In addition to that, if I recall correctly, there shouldn't be any MD5 hashes anymore in META-INF/manifest.mf, only SHA1 and SHA256 hashes.

Yep! We added them to autograph in https://github.com/mozilla-services/autograph/pull/128 then removed them in https://github.com/mozilla-services/autograph/pull/142

**Comment by EnTeQuAk** [comment]
This is now enabled on -dev.

![screenshot from 2019-01-31 09-15-10](https://user-images.githubusercontent.com/139033/52040803-d563e200-2538-11e9-82fc-1b3d6b238bfb.png)

**Comment by AlexandraMoga** [comment]
Here are my test conclusions after having verified this on AMO -dev with FF65:

**New signature applied for all the following cases**:
1.new submissions for all addon types
2.new version submissions
3.unlisted submissions
4.pre-signed packages (SHA1 or SHA256 algorithms) 
5.addons submitted via the API signing endpoint

I'm going to keep monitoring -dev uploads for the remaining of the week, just in case something was missed.

- New META-INF contents:

![image](https://user-images.githubusercontent.com/31961530/52219659-c87f1f80-28a5-11e9-813d-ff9eb49b6226.png)

- all references to `MD5` hashes are gone: example form `manifest.mf`:

![image](https://user-images.githubusercontent.com/31961530/52219896-417e7700-28a6-11e9-9747-0c7580188c87.png)

@EnTeQuAk we are not signing search tools, right?

**Comment by AlexandraMoga** [comment]
Stage testing went well too. For a full list of tests that have been run, please check https://docs.google.com/document/d/1hl1jqm_QZMyPe2BezZIMQgjXO5g2OYxGlCplX3ZVPhM/edit
